proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+definition consists of the params block:
+
+```yaml
+            components:
+            - name: shaper
+              params:
+                inputs:
+                   query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                   documents:
+                      func: contract_docs
+                      params:
+                        docs: documents
+                        delimiter: " "
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+```
+
+The params block consists of the inputs list. Each item in the inputs list is a dictionary with the key being the
+invocation context variable that we want to modify.
+
+In the example above, we have two items in the inputs list. The first item is a dictionary with the key `query` and the
+second item is a dictionary with the key `documents`.
+
+For the query variable, we want to invoke the function `expand` and store the result in the variable `questions`.
+The `expand` function takes two keyword parameters: `expand_target` and `size`. The `expand_target` parameter is the
+name of the variable in the invocation context that we want to expand. The `size`parameter is a result of the `len`
+function invocation on the variable `documents`.
+
+For the documents variable, we want to invoke the function `contract_docs` and store the result in the same variable.
+Therefore, after the invocation, the documents variable will hold a result of `contract_docs` function invocation while
+we'll also have a new variable `questions` in the invocation context. The questions variable will hold a result of
+`expand` function invocation.
+
+The important thing to note here is that we can invoke functions with both keyword and positional parameters. Function
+`len` is an example of a function that takes non-keyword positional parameters. The `contract_docs` and `expand` function
+take keyword parameters. These functions can also be invoked with positional parameters but that is not recommended.
+
+
+### Default parameters
+
+The InputOutputShaper component can also be configured with default parameters. Default parameters are used when we
+don't specify the parameters for a function invocation. The default parameters are specified in the function definition.
+
+For example, in the YAML snippet definition below, we have a function `expand` that takes two keyword parameters:
+`expand_target` and `size`. However, we haven't specified either of these parameters in the YAML config. This is
+possible because we assume that the first parameter is always the variable we want to invoke the function on. In this
```

Comment:
instead of this assumption, the `expand` function and all other shaper functions could have an `in_place` parameter that is set to True by default.